File: frontend/src/api/cultural.ts
import { httpClient } from "./http";

export interface VideoRecommendation {
  id: number;
  title: string;
  description?: string | null;
  platform: string;
  url: string;
  thumbnail_url?: string | null;
  opera_genre?: string | null;
  duration_display?: string | null;
  is_active: boolean;
  created_at: string;
}

export interface WeeklyDigestItem {
  title: string;
  url: string;
  type: "article" | "video";
  desc?: string | null;
}

export interface WeeklyDigest {
  id: number;
  year: number;
  week_number: number;
  title: string;
  summary?: string | null;
  items: WeeklyDigestItem[];
  published_at?: string | null;
  is_published: boolean;
  created_at: string;
}

export interface PaginationMeta { page: number; page_size: number; total: number }

export async function fetchVideoRecommendations(params: { genre?: string; page?: number; pageSize?: number } = {}) {
  const { genre, page = 1, pageSize = 20 } = params;
  const { data } = await httpClient.get<{ data: VideoRecommendation[]; meta: PaginationMeta }>("/cultural/videos", {
    params: { genre, page, page_size: pageSize }
  });
  return data;
}

export async function fetchWeeklyDigests(params: { page?: number; pageSize?: number } = {}) {
  const { page = 1, pageSize = 10 } = params;
  const { data } = await httpClient.get<{ data: WeeklyDigest[]; meta: PaginationMeta }>("/cultural/weekly", {
    params: { page, page_size: pageSize }
  });
  return data;
}

export async function fetchLatestDigest(): Promise<WeeklyDigest | null> {
  const { data } = await httpClient.get<{ data: WeeklyDigest | null }>("/cultural/weekly/latest");
  return data.data;
}

File: backend/app/schemas/cultural.py
"""文化分享 schema。"""
from datetime import datetime
from typing import Any, List, Optional
import json
from pydantic import BaseModel, Field, validator


class VideoRecommendationRead(BaseModel):
    id: int
    title: str
    description: Optional[str]
    platform: str
    url: str
    thumbnail_url: Optional[str]
    opera_genre: Optional[str]
    duration_display: Optional[str]
    is_active: bool
    created_at: datetime

    class Config:
        orm_mode = True


class WeeklyDigestItemRead(BaseModel):
    title: str
    url: str
    type: str = "article"   # article / video
    desc: Optional[str] = None


class WeeklyDigestRead(BaseModel):
    id: int
    year: int
    week_number: int
    title: str
    summary: Optional[str]
    items: List[WeeklyDigestItemRead] = Field(default_factory=list)
    published_at: Optional[datetime]
    is_published: bool
    created_at: datetime

    @classmethod
    def from_orm(cls, obj: Any) -> "WeeklyDigestRead":  # type: ignore[override]
        raw = json.loads(obj.items_json or "[]")
        items = [WeeklyDigestItemRead(**i) for i in raw]
        return cls(
            id=obj.id,
            year=obj.year,
            week_number=obj.week_number,
            title=obj.title,
            summary=obj.summary,
            items=items,
            published_at=obj.published_at,
            is_published=obj.is_published,
            created_at=obj.created_at,
        )

    class Config:
        orm_mode = True
